readthedocs/integrations/models.py

```diff
@@ -140,3 +158,156 @@ def formatted_request_body(self):
     @property
     def formatted_response_body(self):
         return self.formatted_json('response_body')
+
+
+class IntegrationQuerySet(models.QuerySet):
+
+    """Return a subclass of Integration, based on the integration type
+
+    .. note::
+        This doesn't affect queries currently, only fetching of an object
+    """
+
+    def _get_subclass(self, integration_type):
+        # Build a mapping of integration_type -> class dynamically
+        class_map = dict(
+            (cls.integration_type_id, cls)
+            for cls in self.model.__subclasses__()
+            if hasattr(cls, 'integration_type_id')
+        )
+        return class_map.get(integration_type)
+
+    def _get_subclass_replacement(self, original):
+        """Replace model instance on Integration subclasses
+
+        This is based on the ``integration_type`` field, and is used to provide
+        specific functionality to and integration via a proxy subclass of the
+        Integration model.
+        """
+        cls_replace = self._get_subclass(original.integration_type)
+        if cls_replace is None:
+            return original
+        new = cls_replace()
+        for k, v in original.__dict__.items():
+            new.__dict__[k] = v
+        return new
+
+    def get(self, *args, **kwargs):
+        original = super(IntegrationQuerySet, self).get(*args, **kwargs)
+        return self._get_subclass_replacement(original)
+
+    def subclass(self, instance):
+        return self._get_subclass_replacement(instance)
+
+    def create(self, *args, **kwargs):  # pylint: disable=unused-argument
+        """Override of create method to use subclass instance instead
+
+        Instead of using the underlying Integration model to create this
+        instance, we get the correct subclass to use instead. This allows for
+        overrides to ``save`` and other model functions on object creation.
+        """
+        model_cls = self._get_subclass(kwargs.get('integration_type'))
+        if model_cls is None:
+            model_cls = self.model
+        obj = model_cls(**kwargs)
+        self._for_write = True
+        obj.save(force_insert=True, using=self.db)
+        return obj
+
+
+class Integration(models.Model):
+
+    """Inbound webhook integration for projects"""
+
+    GITHUB_WEBHOOK = 'github_webhook'
+    BITBUCKET_WEBHOOK = 'bitbucket_webhook'
+    GITLAB_WEBHOOK = 'gitlab_webhook'
+    API_WEBHOOK = 'api_webhook'
```

Comment:
Should these be defined in a `constants.py` file, instead of on the class directly?